File: src/sim.py
from typing import Dict

from .configs import Config
from .const import SeedExhausted, StockRow, State, Status


def oneday(
    c: StockRow,
    s: State,
    config: Config,
    RSI: Dict[str, float],
    VOLATILITY: Dict[str, float],
    URATE: Dict[str, float],
) -> State:
    rsi = RSI[c.date]
    vol = VOLATILITY[c.date]
    urate = URATE[c.date]

    margin = config.margin
    seed_rate = s.remaining_seed / s.seed
    seed_rate = (
        (seed_rate - config.min_seed_rate) / (1 - config.min_seed_rate)
        if config.min_seed_rate < 1
        else 0
    )
    bullish_rsi = config.bullish_rsi + max(seed_rate, 0) * (
        100 - config.bullish_rsi
    )

    daily_seed: float = s.seed / config.term
    new_s = State.from_(s, c)

    if s.stock_qty > 0 and c.close_price > s.avg_price * (1 + margin):
        new_s.sell(qty=s.stock_qty, sell_price=c.close_price, sold=True)

    else:
        dqtyD = float(daily_seed / c.close_price)
        rate = float(1)

        if dqtyD < 1:
            raise SeedExhausted

        if rsi > bullish_rsi:
            rate = 0

        if (
            urate < config.burst_urate
            and vol < 0
            and abs(vol) > config.burst_vol
        ):
            rate *= (
                1
                + config.burst_scale
                * (abs(vol) - config.burst_vol)
                / config.burst_vol
            )

        dqty = int(dqtyD * rate)
        if s.remaining_seed < c.close_price:
            assert s.status != Status.Sold

            if new_s.cycle_left():  # cycles left # TODO: consider RSI?
                rate = config.sell_base + (
                    config.sell_limit - config.sell_base
                ) * (1 - urate)

                sell_qty = int(s.stock_qty * rate)
                new_s.sell(qty=sell_qty, sell_price=c.close_price)

            else:  # exhausted
                new_s.sell(qty=s.stock_qty, sell_price=c.close_price)

        elif s.remaining_seed >= dqty * c.close_price:
            new_s.buy(qty=dqty, buy_price=c.close_price)

        else:  # s.remaining_seed >= c.close_price
            dqty = int(s.remaining_seed / c.close_price)
            new_s.buy(qty=dqty, buy_price=c.close_price)

    new_s.complete()

    return new_s

File: src/sim.cpp
#include "sim.h"
#include <stdexcept>
#include <cmath>

State oneday(
    const StockRow& c,
    const State& s,
    const Config& config,
    const std::unordered_map<std::string, double>& RSI,
    const std::unordered_map<std::string, double>& VOLATILITY,
    const std::unordered_map<std::string, double>& URATE
) {
    double margin = config.margin;
    
    auto rsi_it = RSI.find(c.date);
    auto vol_it = VOLATILITY.find(c.date);
    auto urate_it = URATE.find(c.date);
    
    if (rsi_it == RSI.end() || vol_it == VOLATILITY.end() || urate_it == URATE.end()) {
        throw std::runtime_error("Missing data for date: " + c.date);
    }
    
    double rsi = rsi_it->second;
    double vol = vol_it->second;
    double urate = urate_it->second;

    double daily_seed = s.seed / config.term;
    State new_s = State::from(s, c);

    if (s.stock_qty > 0 && c.close_price > s.avg_price * (1 + margin)) {
        new_s.sell(static_cast<int>(s.stock_qty), c.close_price, true);
    } else {
        double dqtyD = daily_seed / c.close_price;
        double rate = 1.0;

        if (dqtyD < 1) {
            throw SeedExhausted();
        }

        if (rsi > config.bullish_rsi) {
            rate = 0;
        }

        if (urate < config.burst_urate && vol < 0 && std::abs(vol) > config.burst_vol) {
            rate *= (1 + config.burst_scale * (std::abs(vol) - config.burst_vol) / config.burst_vol);
        }

        int dqty = static_cast<int>(dqtyD * rate);
        
        if (s.remaining_seed < c.close_price) {
            // assert s.status != Status::Sold

            if (new_s.cycle_left()) {
                double sell_rate = config.sell_base + (config.sell_limit - config.sell_base) * (1 - urate);
                int sell_qty = static_cast<int>(s.stock_qty * sell_rate);
                new_s.sell(sell_qty, c.close_price);
            } else {
                new_s.sell(static_cast<int>(s.stock_qty), c.close_price);
            }
        } else if (s.remaining_seed >= dqty * c.close_price) {
            new_s.buy(dqty, c.close_price);
        } else {
            dqty = static_cast<int>(s.remaining_seed / c.close_price);
            new_s.buy(dqty, c.close_price);
        }
    }

    new_s.complete();
    return new_s;
}